log_tcp/log_tcp/hosts_access.5
.TH HOSTS_ACCESS 5
.ad
.fi
.SH NAME
hosts_access \- host access control files
.SH DESCRIPTION
.ad
.fi
This manual page describes a simple, but effective, access control
facility that is based on host (or domain) names, netgroups, internet
addresses (or network numbers) and on network daemon process names.
.PP
In the following text, \fIdaemon\fP is the the process name (argv[0]
value) of a network daemon process, and \fIclient\fP is the name of
a remote host (or its internet address if the name is not available).
.IP o
Access will be granted when a (daemon,client) pair is matched by an
entry in the \fI/etc/hosts.allow\fP file.
.IP o
If the previous test fails (perhaps because the \fIhosts.allow\fP file
does not exist), access will be denied when a (daemon,client) pair is
matched by an entry in the \fI/etc/hosts.deny\fP file.
.IP o
If the previous test fails (perhaps because the \fIhosts.deny\fP file
does not exist), access will be granted.
.PP
A non-existing access control file is treated as if it were an empty
file. Thus, access control can be turned off by providing no access
control files.
.PP
The format of the access control files is as follows.
.IP o
Lines that begin with a `#\' character are ignored.
.IP o
Other lines should have the format (things between [] are optional):
.sp
.ti +3
daemon_list : client_list [ : shell_command ]
.PP
\fIdaemon_list\fP is a list of one or more daemon process names
(argv[0] values).  \fIclient_list\fP is a list of one or more host
names, domain names, netgroups, internet addresses or internet network
numbers.  List elements should be separated by blanks and/or commas.
With the exception of netgroup lookups, all access control lookups are
case insensitive.
.PP
Client_list fields that specify a domain name should begin with a
`.\' character (see example below). Internet \fInetwork\fP numbers (as
opposed to internet \fIhost\fP numbers) should be terminated with a `.\'
character. A netgroup name should begin with the `@\' character.
Netgroups are usually supported on systems with NIS (formerly YP)
data bases.
.PP
Special meaning is given to the magic token \fIALL\fP.  If it appears
in a daemon_list, this token matches all network daemon process
names.  If the magic token appears in a client_list, it matches
all clients.  Another token that receives special treatment is
\fILOCAL\fP. If it appears in a daemon_list or client_list,
it matches any string that does not contain a dot character.
.PP
The optional \fIshell_command\fP is executed if the table entry is the
first one that matches the (daemon,client) pair. The format is: one or
more statements separated by a `;\' character. Prior to execution,
the following substitutions are done on the shell_command:
.TP
%h
is replaced by the remote host name
(or address, if the host name is not available).
.TP
%d
is replaced by the daemon process name (argv[0] value).
.TP
%p
is replaced by the daemon process id.
.TP
%%
is replaced by a single `%\' character.
.PP
The resulting command is given to \fI/bin/sh\fP, with standard input,
output and error connected to \fI/dev/null\fP.  Specify an `&\' at the
end of the command if you do not want to wait until it has completed.
.PP
Shell_commands should not rely on the PATH setting of the inetd.
Instead, they should use absolute path names, or they should begin with
an explicit PATH=whatever statement.
.SH EXAMPLES
The following example restricts all services to hosts within the local
domain (no `.\' character in the host name), all hosts below 
the \fI.some.domain\fP, and all hosts in the \fIthatgroup\fP netgroup:
.PP
/etc/hosts.allow: 
.in +3
ALL: LOCAL, .some.domain, @thatgroup
.PP
/etc/hosts.deny: 
.in +3
ALL: ALL
.PP
In order to deny some hosts all services, except ftp:
.PP
/etc/hosts.allow: 
.in +3
in.ftpd: ALL
.PP
/etc/hosts.deny: 
.in +3
ALL: some.host.name, .some.domain
.PP
The following example implements a backfinger facility that is
triggered when someone invokes your tftp service from a host outside the
local domain:
.PP
/etc/hosts.allow:
.in +3
.nf
in.tftpd: LOCAL, .my.domain
.PP
/etc/hosts.deny:
.in +3
.nf
in.tftpd: ALL: (/bin/date; /usr/ucb/finger -l @%h) >>/var/adm/%d.log &
.fi
.PP
If your tftp daemon is run under a uid other than root (strongly
recommended), the /var/adm/whatever.log file should be writable for
that uid. Watch out for infinite backfinger loops!
.SH DIAGNOSTICS
.ad
.fi
An error is reported when a syntax error is found in a host access
control file; when the length of an entry in a host access control file
exceeds the STDIO buffer size; when the result of %<character>
expansion would overflow an internal buffer; when a system call fails
that shouldn\'t.
.SH FILES
.na
.nf
/etc/hosts.allow, (daemon,client) pairs that are granted access.
/etc/hosts.deny, (daemon,client) pairs that are denied access.
.SH BUGS
.ad
.fi
If there are problems with a name server, the access control software
will use a host\'s address instead of its name.  A workaround is to
also list internet addresses and network numbers in the access-control
files.
.PP
Domain name server lookups are case insensitive; NIS (formerly YP)
netgroup lookups are case sensitive.
.PP
Some UDP (and RPC) daemons linger around for a while after they have
finished their work, just in case another request comes in (in the
inetd configuration file, these daemons are registered with the `wait\'
option).  In such cases, access control will apply only to the request
that started such a daemon. This restriction usually does not apply to
connection-oriented (TCP) services.
.SH AUTHOR
.na
.nf
Wietse Venema
Eindhoven University of Technology
Department of Mathematics and Computer Science
Den Dolech 2, P.O. Box 513, 5600 MB Eindhoven, The Netherlands
\" @(#) hosts_access.5 1.8 91/10/02 23:02:00


log_tcp/log_tcp/hosts_access.c
#ifdef HOSTS_ACCESS

 /*
  * This module implements a simple but effective form of access control
  * based on host (or domain) names, netgroup, internet addresses (or network
  * numbers) and daemon process names, with wild card support. Upon the first
  * match with an entry in the access-control tables, an optional shell
  * command is executed.
  * 
  * Diagnostics are reported through syslog(3).
  * 
  * Compile with -DHOSTS_ACCESS in order to enable access control. See the
  * hosts_access(5) manual page for details.
  * 
  * Compile with -DNETGROUP if your library provides support for netgroups.
  * 
  * Author: Wietse Venema, Eindhoven University of Technology, The Netherlands.
  */

#ifndef lint
static char sccsid[] = "@(#) hosts_access.c 1.6 91/10/02 23:01:49";
#endif

 /* System libraries. */

#include <sys/types.h>
#include <sys/param.h>
#include <stdio.h>
#include <syslog.h>
#include <ctype.h>

extern char *fgets();
extern char *strchr();
extern char *strtok();
extern void exit();

/* Local stuff. */

#include "log_tcp.h"

/* Path names of the access control files. */

#define HOSTS_ALLOW	"/etc/hosts.allow"
#define HOSTS_DENY	"/etc/hosts.deny"

/* Delimiters for lists of daemons or clients. */

static char sep[] = ", \t";

/* Constants to be used in assignments only, not in comparisons... */

#define	YES		1
#define	NO		0

/* Forward declarations. */

static int table_match();
static int list_match();

/* hosts_access - host access control facility */

int hosts_access(daemon, client)
char   *daemon;
char   *client;
{

    /*
     * If the (daemon, client) pair is matched by an entry in the file
     * /etc/hosts.allow, access is granted. Otherwise, if the (daemon,
     * client) pair is matched by an entry in the file /etc/hosts.deny,
     * access is denied. Otherwise, access is granted. A non-existent
     * access-control file is treated as an empty file.
     */

    if (table_match(HOSTS_ALLOW, daemon, client))
	return (YES);
    if (table_match(HOSTS_DENY, daemon, client)) 
	return (NO);
    return (YES);
}

/* table_match - match table entries with (daemon, client) pair */

static int table_match(table, daemon, client)
char   *table;
char   *daemon;
char   *client;
{
    FILE   *fp;
    char    sv_list[BUFSIZ];		/* becomes list of daemons */
    char   *cl_list;			/* becomes list of clients */
    char   *sh_cmd;			/* becomes optional shell command */
    int     match = NO;
    int     end;

    /*
     * Process the table one line at a time. Lines that begin with a '#'
     * character are ignored. Non-comment lines are broken at the ':'
     * character (we complain if there is none). The first field is matched
     * against the daemon process name (argv[0]), the second field against
     * the host name. A non-existing table is treated as if it were an empty
     * table. The optional shell command (third field) is executed when the
     * first match is found.
     */

    if (fp = fopen(table, "r")) {
	while (match == 0 && fgets(sv_list, sizeof(sv_list), fp)) {
	    if (sv_list[end = strlen(sv_list) - 1] != '\n') {
		syslog(LOG_ERR, "%s: line exceeds STDIO buffer size", table);
		continue;
	    } else {
		sv_list[end] = '\0';		/* strip trailing newline */
	    }
	    if (sv_list[0] == '#') {		/* skip comments */
		continue;
	    } else if ((cl_list = strchr(sv_list, ':')) == 0) {
		syslog(LOG_ERR, "%s: malformed entry: \"%s\"", table, sv_list);
		continue;
	    } else {
		*cl_list++ = '\0';		/* split 1st and 2nd fields */
		if ((sh_cmd = strchr(cl_list, ':')) != 0)
		    *sh_cmd++ = '\0';		/* split 2nd and 3rd fields */
		match = (list_match(sv_list, daemon)
			 && list_match(cl_list, client));
	    }
	}
	(void) fclose(fp);
    }
    if (match && sh_cmd)
	shell_cmd(sh_cmd, daemon, client);
    return (match);
}

/* list_match - match a string against a list of tokens */

static int list_match(list, string)
char   *list;
char   *string;
{
    char   *tok;
    int     tok_len;
    int     str_len;

    /*
     * Process tokens one at a time. If a token has the magic value "ALL" the
     * match always succeeds. If the token is a domain name, return YES if it
     * matches the last fields of the string. If the token has the magic
     * value "LOCAL", return YES if the string does not contain a "."
     * character. If the token is a network number, return YES if it matches
     * the head of the string. If the token looks like a netgroup name,
     * return YES if the string is a (host) member of the netgroup.
     * Otherwise, return YES if the token fully matches the string. Note: we
     * assume that a daemon process name never begins or ends with a "." or
     * "@" character.
     */

    for (tok = strtok(list, sep); tok; tok = strtok((char *) 0, sep)) {
	if (tok[0] == '.') {			/* domain: match last fields */
	    if ((str_len = strlen(string)) > (tok_len = strlen(tok))
		&& strcasecmp(tok, string + str_len - tok_len) == 0)
		return (YES);
#ifdef	NETGROUP
	} else if (tok[0] == '@') {		/* netgroup: look it up */
	    if (innetgr(tok + 1, string, (char *) 0, (char *) 0))
		return (YES);
#endif
	} else if (strcasecmp(tok, "ALL") == 0) {	/* all: match any */
	    return (YES);
	} else if (strcasecmp(tok, "LOCAL") == 0) {	/* local: no dots */
	    if (strchr(string, '.') == 0)
		return (YES);
	} else if (!strcasecmp(tok, string)) {	/* match host name or address */
	    return (YES);
	} else if (tok[(tok_len = strlen(tok)) - 1] == '.'	/* net number */
		   && strncmp(tok, string, tok_len) == 0) {
	    return (YES);
	}
    }
    return (NO);
}

#endif
